src/predictor/gpu_predictor.cu

```diff
@@ -510,6 +563,50 @@ class GPUPredictor : public xgboost::Predictor {
     }
   }
 
+  std::vector<gpu_treeshap::PathElement> ExtractPaths(
+      const gbm::GBTreeModel& model, size_t tree_limit) {
+    std::vector<gpu_treeshap::PathElement> paths;
+    size_t path_idx = 0;
+    CHECK_LE(tree_limit, model.trees.size());
+    for (auto i = 0ull; i < tree_limit; i++) {
+      const auto& tree = *model.trees.at(i);
+      size_t group = model.tree_info[i];
+      for (auto j = 0ull; j < tree.GetNodes().size(); j++) {
+        const auto& nodes = tree.GetNodes();
+        if (nodes[j].IsLeaf() && !nodes[j].IsDeleted()) {
+          auto child = nodes[j];
+          float v = child.LeafValue();
+          size_t child_idx = j;
+          const float inf = std::numeric_limits<float>::infinity();
+          while (!child.IsRoot()) {
+            float child_cover = tree.Stat(child_idx).sum_hess;
+            float parent_cover = tree.Stat(child.Parent()).sum_hess;
+            float zero_fraction = child_cover / parent_cover;
+            CHECK(zero_fraction >= 0.0 && zero_fraction <= 1.0);
+            auto parent = nodes[child.Parent()];
+            CHECK(parent.LeftChild() == child_idx ||
+                  parent.RightChild() == child_idx);
+            bool is_left_path = parent.LeftChild() == child_idx;
+            bool is_missing_path = (!parent.DefaultLeft() && !is_left_path) ||
+                                   (parent.DefaultLeft() && is_left_path);
+            float lower_bound = is_left_path ? -inf : parent.SplitCond();
+            float upper_bound = is_left_path ? parent.SplitCond() : inf;
+            paths.emplace_back(path_idx, parent.SplitIndex(), group,
+
+                               lower_bound, upper_bound, is_missing_path,
```

Comment:
@RAMitchell  Is the feature bounds  (`lower_bound`, `upper_bound`) used for detecting duplicate features in each tree path? I took a brief look at GPUTreeshap, and that's what I inferred those bounds to be for. It's pretty cool that we don't have to "unwind" paths as in the original SHAP algorithm.

Reminder to myself: Make time to study GPUTreeshap code.

Reply:
Also, does a unique traversal path in a tree consist of a sequence of `gpu_treeshap::PathElement`?

Reply:
Yes, any number of splits on the same feature can be represented by these bounds.

> Also, does a unique traversal path in a tree consist of a sequence of gpu_treeshap::PathElement?

Yes